### DISCUSSION TITLE
Interact between commands and Haspmota

### DISCUSSION THREAD

**Discussion by valdus2** [arendst/Tasmota#21220] [Q&A] [answered]
Is there a way to interaract between a command given from console or web etc. with Haspmota ?
Changing for example a page with some http get command or running a berry script, which is called again from web, 
which in turn will change the page in Haspmota or whatever.

Thank you.

**Comment by sfromis** [comment]
In general, you can change HASPmota objects from Berry, using assignments like
`p1b2.text = str(temperature)`

Such an assignment can obviously be part of a command you define in Berry, to be executed by one of the multiple ways you can issue Tasmota commands. You could also use the existing `Br` command to run a snippet of Berry code, including an assignment, like
`Br p1b2.text = '12.4'`

Depending on the data source, using MQTT is also a good way to send data to Tasmota, with Berry being well-integrated with both MQTT and HASPmota. Via MQTT, you can also have HASPmota objects updated without having any Berry code at all.

**Reply by phoenix83** [reply]
"Via MQTT, you can also have HASPmota objects updated without having any Berry code at all."

Excuse my ignorance, but I can't find any info on this. Are you saying there is a way (without the Br command or custom commands) to send HASPmota updates via MQTT?

**Reply by sfromis** [reply]
I was not trying to suggest "without any command", but instead being able to update values without already having a predefined chunk of Berry code to pick up values, and transfer them to HASPmota. 

Meaning that you can have a device with HASPmota and zero Berry code of your own, and use MQTT to send values, where the `Br` command is a simple and convenient way.

That is if you do not want the more robust style of encapsulating the HASPmota handling to avoid having internal structure dependencies (like the specific global variable names for HASPmota objects) being spread outside of the Tasmota device.

**Reply by phoenix83** [reply]
Ah ok, thanks for the clarification.

**Comment by valdus2** [comment]
I think you misunderstood me...I mean using a simple command from console to change page 
as we have the command for example `power 1` and the relay opens...

Is there a command such as `changepage 2` which will turn in page 2 in haspmota?
Or a command such as `runberry filename.be` which will run a script ?

Thank you.

**Reply by sfromis** [reply, marked as answer]
No, I did understand what you asked for, and the `Br` command I detailed can be executed like any other command, like `Power`, etc. I even posted a full example.

Or it is simple to [create a command](https://tasmota.github.io/docs/Berry-Cookbook/#adding-commands-to-tasmota) in Berry, like
`tasmota.add_cmd('Changepage', ....)

You can also use the `Br` command to execute a file of Berry code via the `load` function.
`Br load('filename.be')`

**Comment by valdus2** [comment]
Ok, now I understood you...

So I can do this: `On power#state 1 do Br load('filename.be') endon` ?

**Reply by sfromis** [reply]
Yes. If you have some function already loaded in Berry, you can also `Br myfunction(%value%)` (this only works in a "classic" rule).

You can also do such in Berry without complicating things with "classic" rules.

```
tasmota.add_rule('Power#state', 
  def (value,trigger,payload)
    p1b2=value    # or whatever
  end)
```

**Comment by valdus2** [comment]
Thank you sfromis....Helpful as always.
